src/utils/utils.py

```diff
@@ -368,552 +368,6 @@ def split_data(data, ratio, shuffle=1):
-    """
-    For use with :func:`combine_tensors`.  This function computes the resultant dimension when
-    calling ``combine_tensors(combination, tensors)``, when the tensor dimension is known.  This is
-    necessary for knowing the sizes of weight matrices when building models that use
-    ``combine_tensors``.
-
-    Parameters
-    ----------
-    combination : ``str``
-        A comma-separated list of combination pieces, like ``"1,2,1*2"``, specified identically to
-        ``combination`` in :func:`combine_tensors`.
-    tensor_dims : ``List[int]``
-        A list of tensor dimensions, where each dimension is from the `last axis` of the tensors
-        that will be input to :func:`combine_tensors`.
-    """
-    if len(tensor_dims) > 9:
-        raise ConfigurationError("Double-digit tensor lists not currently supported")
-    combination = combination.replace('x', '1').replace('y', '2')
-    return sum([_get_combination_dim(piece, tensor_dims) for piece in combination.split(',')])
-
-
-def _get_combination_dim(combination: str, tensor_dims: List[int]) -> int:
-    if combination.isdigit():
-        index = int(combination) - 1
-        return tensor_dims[index]
-    else:
-        if len(combination) != 3:
-            raise ConfigurationError("Invalid combination: " + combination)
-        first_tensor_dim = _get_combination_dim(combination[0], tensor_dims)
-        second_tensor_dim = _get_combination_dim(combination[2], tensor_dims)
-        operation = combination[1]
-        if first_tensor_dim != second_tensor_dim:
-            raise ConfigurationError(
-                "Tensor dims must match for operation \"{}\"".format(operation))
-        return first_tensor_dim
-
-
-def logsumexp(tensor: torch.Tensor,
-              dim: int = -1,
-              keepdim: bool = False) -> torch.Tensor:
-    """
-    A numerically stable computation of logsumexp. This is mathematically equivalent to
-    `tensor.exp().sum(dim, keep=keepdim).log()`.  This function is typically used for summing log
-    probabilities.
-
-    Parameters
-    ----------
-    tensor : torch.FloatTensor, required.
-        A tensor of arbitrary size.
-    dim : int, optional (default = -1)
-        The dimension of the tensor to apply the logsumexp to.
-    keepdim: bool, optional (default = False)
-        Whether to retain a dimension of size one at the dimension we reduce over.
-    """
-    max_score, _ = tensor.max(dim, keepdim=keepdim)
-    if keepdim:
-        stable_vec = tensor - max_score
-    else:
-        stable_vec = tensor - max_score.unsqueeze(dim)
-    return max_score + (stable_vec.exp().sum(dim, keepdim=keepdim)).log()
-
-
-def subsequent_mask(size):
-    "Mask out subsequent positions."
-    attn_shape = (1, size, size)
-    subsequent_mask = np.triu(np.ones(attn_shape), k=1).astype('uint8')
-    return torch.from_numpy(subsequent_mask) == 0
-
-
 @Seq2SeqEncoder.register("masked_multi_head_self_attention")
```

Comment:
TODO: We should probably move this (and anything like it) to something under `src/modules`